interval/search_tree.go

```diff
@@ -113,6 +116,46 @@ func (st *SearchTree[V, T]) IsEmpty() bool {
 	return st.root == nil
 }
 
+// GobEncode encodes the tree (compatible with encoding/gob).
+func (st *SearchTree[V, T]) GobEncode() ([]byte, error) {
+	var b bytes.Buffer
+	enc := gob.NewEncoder(&b)
+
+	if err := enc.Encode(st.config.allowIntervalPoint); err != nil {
+		return nil, err
+	}
+
+	if st.root != nil {
+		if err := enc.Encode(st.root); err != nil {
+			return nil, err
+		}
+	}
+
+	return b.Bytes(), nil
+}
+
+// GobDecode decodes the tree (compatible with encoding/gob).
+func (st *SearchTree[V, T]) GobDecode(data []byte) error {
+	b := bytes.NewBuffer(data)
+	enc := gob.NewDecoder(b)
+
+	if err := enc.Decode(&st.config.allowIntervalPoint); err != nil {
+		return err
+	}
+
+	if err := enc.Decode(&st.root); err != nil {
+		if err != io.EOF {
+			return err
+		}
```

Comment:
What do you think of adding a test for this code branch to guarantee we are properly handling non `io.EOF` errors?

For example:
```go
func TestMultiValueSearchTree_DecodingWithError(t *testing.T) {
	orig := NewMultiValueSearchTreeWithOptions[string, int](func(x, y int) int { return x - y }, TreeWithIntervalPoint())
	orig.Insert(17, 19, "node1")
	orig.Insert(5, 8, "node2")
	orig.Insert(21, 24, "node3")
	orig.Insert(21, 24, "node4")
	orig.Insert(4, 4, "node5")

	dest := NewMultiValueSearchTree[string, string](func(x, y string) int { return strings.Compare(x, y) })

	b := encodeMultiValueTree(t, orig)
	err := decodeMultiValueTreeErr(dest, b)
	if err == nil {
		t.Fatal("got unexpected <nil> error; want not nil")
	}
}
```

Note the error occurs because the trees `orig` and `dest` have different key types, `int` and `string`, respectively.

Reply:
Done: [Add tests for decoding errors](https://github.com/rdleal/intervalst/pull/9/commits/d1d509cb3b54ee67cc1e7460278ebb4cfc868420)

Empty trees don't generate errors, so I didn't include those cases.